src/key.cpp

```diff
@@ -172,6 +174,62 @@ bool CKey::Negate()
     return secp256k1_ec_seckey_negate(secp256k1_context_sign, keydata.data());
 }
 
+void CKey::TweakAdd(const unsigned char *tweak32)
+{
+    assert(fValid);
+    // Modify the current CKey's data directly.
+    int ret = secp256k1_ec_seckey_tweak_add(secp256k1_context_sign, keydata.data(), tweak32);
+    assert(ret);
+}
+
+void CKey::TweakMultiply(const unsigned char *tweak32)
+{
+    assert(fValid);
+
+    // Modify the current CKey's data directly.
+    int ret = secp256k1_ec_seckey_tweak_mul(secp256k1_context_sign, keydata.data(), tweak32);
+    assert(ret);
+}
+
+// WARNING: DO NOT USE THIS FUNCTION OUTSIDE OF SILENT PAYMENTS
+// This function returns the un-hashed ECDH result and is specific to silent payments.
+// If ECDH is required for a different use-case, use a function which hashes the result (as is standard practice).
+CPubKey CKey::UnhashedECDH(const CPubKey& pubkey) const
+{
+    unsigned char shared_secret[33];
+
+    secp256k1_pubkey ecdh_pubkey;
```

Comment:
Hm, it seems a bit odd to assert that the passed pubkey is valid? I think it's safe to assume that validation was done upstream by the caller. Also, I agree that we want to avoid doing expensive checks here as this function is called during scanning.

I'll add the assert for the private key, though, as this seems in keeping with what the other code with private keys is doing.

Reply:
Added `assert(fValid)` check for CKey and also made the functions consistent with the rest of the CPubKey/CKey functions by returning a bool instead of an `assert` inside a void function